File: codetraverse/ast_diff/rustdiff.py
import json
from tree_sitter import Language, Parser, Node
import tree_sitter_rust
from .Detailedchanges import DetailedChanges
from .basefilediff import BaseFileDiff
class RustFileDiff(BaseFileDiff):
    """Analyzes and compares two Rust ASTs for semantic differences."""
    def __init__(self, module_name=""):
        self.changes = DetailedChanges(module_name)

    def get_decl_name(self, node: Node) -> str:
        """Finds the name of a Rust declaration."""
        # Handle `impl` blocks by creating a composite name
        if node.type == 'impl_item':
            trait = node.child_by_field_name('trait')
            type_node = node.child_by_field_name('type')
            if trait and type_node:
                return f"{trait.text.decode('utf8')} for {type_node.text.decode('utf8')}"
            elif type_node:
                return type_node.text.decode('utf8')
        
        # Handle `use` declarations
        if node.type == 'use_declaration':
            arg_node = node.child_by_field_name('argument')
            if arg_node:
                return arg_node.text.decode('utf8')

        # For most other declarations, the name is in a 'name' field
        name_node = node.child_by_field_name('name')
        if name_node:
            return name_node.text.decode('utf8')
        return None

    def extract_components(self, root: Node):
        """Extracts all top-level declarations from a Rust AST."""
        items = {
            "functions": {}, "structs": {}, "enums": {}, "traits": {},
            "impls": {}, "uses": {}, "consts": {},
        }
        
        node_type_map = {
            "function_item": items["functions"],
            "struct_item": items["structs"],
            "enum_item": items["enums"],
            "trait_item": items["traits"],
            "impl_item": items["impls"],
            "use_declaration": items["uses"],
            "const_item": items["consts"],
            "static_item": items["consts"], # Treat statics like consts
            "type_item": items["structs"], # Treat type aliases like structs
        }
        
        if root.type == 'source_file':
            for child in root.children:
                if child.type in node_type_map:
                    name = self.get_decl_name(child)
                    if name:
                        target_dict = node_type_map[child.type]
                        target_dict[name] = (child, child.text.decode('utf8'), child.start_point, child.end_point)

        return items

    def diff_components(self, before_map: dict, after_map: dict):
        """Compares two dictionaries of components and returns the diff."""
        before_names = set(before_map.keys())
        after_names = set(after_map.keys())
        added_names = after_names - before_names
        deleted_names = before_names - after_names
        common_names = before_names & after_names
        
        added = [(n, after_map[n][1], {"start": after_map[n][2], "end": after_map[n][3]}) for n in sorted(added_names)]
        deleted = [(n, before_map[n][1], {"start": before_map[n][2], "end": before_map[n][3]}) for n in sorted(deleted_names)]
        modified = []
        for name in sorted(common_names):
            _, old_body, old_start, old_end = before_map[name]
            _, new_body, new_start, new_end = after_map[name]
            if old_body.strip() != new_body.strip():
                 modified.append((name, old_body, new_body, {"old_start": old_start, "old_end": old_end, "new_start": new_start, "new_end": new_end}))
        return {"added": added, "deleted": deleted, "modified": modified}

    def compare_two_files(self, old_file_ast: Node, new_file_ast: Node) -> DetailedChanges:
        """The main method to compare two Rust files."""
        old_items = self.extract_components(old_file_ast.root_node)
        new_items = self.extract_components(new_file_ast.root_node)

        for category in ["functions", "structs", "enums", "traits", "impls", "uses", "consts"]:
            diff = self.diff_components(old_items[category], new_items[category])
            for change_type in ["added", "deleted", "modified"]:
                for data in diff[change_type]:
                    self.changes.add_change(category, change_type, data)

        return self.changes
    
    def process_single_file(self, file_ast: Node, mode="deleted") -> DetailedChanges:
        """Processes a single file that was either entirely added or deleted."""
        items = self.extract_components(file_ast.root_node)

        for category, component_map in items.items():
            for name, data_tuple in component_map.items():
                # data_tuple is (node, text, start_point, end_point)
                item = (name, data_tuple[1], {"start": data_tuple[2], "end": data_tuple[3]})
                self.changes.add_change(category, mode, item)
        
        return self.changes

File: codetraverse/ast_diff/rustdiff.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RustFileDiff = void 0;
var Detailedchanges_1 = require("./Detailedchanges");
var basefilediff_1 = require("./basefilediff");
var RustFileDiff = /** @class */ (function (_super) {
    __extends(RustFileDiff, _super);
    function RustFileDiff(moduleName) {
        if (moduleName === void 0) { moduleName = ""; }
        var _this = _super.call(this, moduleName) || this;
        _this.changes = new Detailedchanges_1.DetailedChanges(moduleName);
        return _this;
    }
    RustFileDiff.prototype.getDeclName = function (node) {
        if (node.type === 'impl_item') {
            var traitNode = node.childForFieldName('trait');
            var typeNode = node.childForFieldName('type');
            if (traitNode && typeNode) {
                return "".concat(traitNode.text, " for ").concat(typeNode.text);
            }
            else if (typeNode) {
                return typeNode.text;
            }
        }
        if (node.type === 'use_declaration') {
            var argNode = node.childForFieldName('argument');
            if (argNode) {
                return argNode.text;
            }
        }
        var nameNode = node.childForFieldName('name');
        if (nameNode) {
            return nameNode.text;
        }
        return null;
    };
    RustFileDiff.prototype.extractComponents = function (root) {
        var items = {
            "functions": {}, "structs": {}, "enums": {}, "traits": {},
            "impls": {}, "uses": {}, "consts": {},
        };
        var nodeTypeMap = {
            "function_item": items.functions,
            "struct_item": items.structs,
            "enum_item": items.enums,
            "trait_item": items.traits,
            "impl_item": items.impls,
            "use_declaration": items.uses,
            "const_item": items.consts,
            "static_item": items.consts,
            "type_item": items.structs,
        };
        if (root.type === 'source_file') {
            for (var _i = 0, _a = root.children; _i < _a.length; _i++) {
                var child = _a[_i];
                if (child.type in nodeTypeMap) {
                    var name_1 = this.getDeclName(child);
                    if (name_1) {
                        var targetDict = nodeTypeMap[child.type];
                        targetDict[name_1] = [child, child.text, child.startPosition, child.endPosition];
                    }
                }
            }
        }
        return items;
    };
    RustFileDiff.prototype.diffComponents = function (beforeMap, afterMap) {
        var beforeNames = new Set(Object.keys(beforeMap));
        var afterNames = new Set(Object.keys(afterMap));
        var addedNames = Array.from(afterNames).filter(function (name) { return !beforeNames.has(name); });
        var deletedNames = Array.from(beforeNames).filter(function (name) { return !afterNames.has(name); });
        var commonNames = Array.from(beforeNames).filter(function (name) { return afterNames.has(name); });
        var added = addedNames.sort().map(function (n) { return [n, afterMap[n][1], { start: afterMap[n][2], end: afterMap[n][3] }]; });
        var deleted = deletedNames.sort().map(function (n) { return [n, beforeMap[n][1], { start: beforeMap[n][2], end: beforeMap[n][3] }]; });
        var modified = [];
        for (var _i = 0, _a = commonNames.sort(); _i < _a.length; _i++) {
            var name_2 = _a[_i];
            var _b = beforeMap[name_2], oldAst = _b[0], oldBody = _b[1], oldStart = _b[2], oldEnd = _b[3];
            var _c = afterMap[name_2], newAst = _c[0], newBody = _c[1], newStart = _c[2], newEnd = _c[3];
            if (oldBody.trim() !== newBody.trim()) {
                modified.push([name_2, oldBody, newBody, { old_start: oldStart, old_end: oldEnd, new_start: newStart, new_end: newEnd }]);
            }
        }
        return { added: added, deleted: deleted, modified: modified };
    };
    RustFileDiff.prototype.compareTwoFiles = function (oldFileAst, newFileAst) {
        var oldItems = this.extractComponents(oldFileAst);
        var newItems = this.extractComponents(newFileAst);
        for (var _i = 0, _a = ["functions", "structs", "enums", "traits", "impls", "uses", "consts"]; _i < _a.length; _i++) {
            var category = _a[_i];
            var diff = this.diffComponents(oldItems[category], newItems[category]);
            for (var changeType in diff) {
                for (var _b = 0, _c = diff[changeType]; _b < _c.length; _b++) {
                    var data = _c[_b];
                    this.changes.add_change(category, changeType, data);
                }
            }
        }
        return this.changes;
    };
    RustFileDiff.prototype.processSingleFile = function (fileAst, mode) {
        if (mode === void 0) { mode = "deleted"; }
        var items = this.extractComponents(fileAst);
        for (var category in items) {
            var componentMap = items[category];
            for (var name_3 in componentMap) {
                var dataTuple = componentMap[name_3];
                var item = [name_3, dataTuple[1], { start: dataTuple[2], end: dataTuple[3] }];
                this.changes.add_change(category, mode, item);
            }
        }
        return this.changes;
    };
    return RustFileDiff;
}(basefilediff_1.BaseFileDiff));
exports.RustFileDiff = RustFileDiff;
